editor/src/components/canvas/controls/select-mode/select-mode-hooks.ts

```diff
@@ -0,0 +1,407 @@
+            event.metaKey,
+            duplicateNewUIDs,
+            start,
+            componentMetadata,
+            moveTargets,
+          ),
+        ),
+      ])
+    },
+    [dispatch, entireEditorStoreRef],
+  )
+}
+
+function callbackAfterDragExceedsThreshold(
+  startEvent: MouseEvent,
+  threshold: number,
+  callback: (event: MouseEvent) => void,
+) {
+  const startPoint = windowPoint(point(startEvent.clientX, startEvent.clientY))
+  function onMouseMove(event: MouseEvent) {
+    if (distance(startPoint, windowPoint(point(event.clientX, event.clientY))) > threshold) {
+      callback(event)
+      removeListeners()
+    }
+  }
+
+  function onMouseUp() {
+    removeListeners()
+  }
+
+  function removeListeners() {
+    window.removeEventListener('mousemove', onMouseMove)
+    window.removeEventListener('mouseup', onMouseUp)
+  }
+
+  window.addEventListener('mousemove', onMouseMove)
+  window.addEventListener('mouseup', onMouseUp)
+}
+
+export function useSelectModeSelectAndHover(): {
+  onMouseOver: (event: React.MouseEvent<HTMLDivElement>) => void
+  onMouseOut: () => void
+  onMouseDown: (event: React.MouseEvent<HTMLDivElement>) => void
+} {
+  const dispatch = useEditorState((store) => store.dispatch, 'useSelectAndHover dispatch')
+  const { maybeHighlightOnHover, maybeClearHighlightsOnHoverEnd } = useMaybeHighlightElement()
+  const findValidTarget = useFindValidTarget()
+  const startDragState = useStartDragState()
+  const windowToCanvasCoordinates = useWindowToCanvasCoordinates()
+
+  const onMouseOver = React.useCallback(
+    (event: React.MouseEvent<HTMLDivElement>) => {
+      const validTemplatePath = findValidTarget(event.target as HTMLDivElement, event.metaKey)
+      if (
+        validTemplatePath != null &&
+        validTemplatePath.selectionMode === 'singleclick' && // we only show highlights for single-click selectable elements
+        !validTemplatePath.isSelected // do not highlight selected elements
+      ) {
+        maybeHighlightOnHover(validTemplatePath.templatePath)
+      }
+    },
+    [maybeHighlightOnHover, findValidTarget],
+  )
+
+  const onMouseOut = React.useCallback(() => {
+    maybeClearHighlightsOnHoverEnd()
+  }, [maybeClearHighlightsOnHoverEnd])
+
+  const onMouseDown = React.useCallback(
+    (event: React.MouseEvent<HTMLDivElement>) => {
+      const foundTarget = findValidTarget(event.target as HTMLDivElement, event.metaKey)
+      if (foundTarget != null) {
+        callbackAfterDragExceedsThreshold(
+          event.nativeEvent,
+          DRAG_START_TRESHOLD,
+          startDragState(
+            foundTarget.templatePath,
+            windowToCanvasCoordinates(windowPoint(point(event.clientX, event.clientY)))
+              .canvasPositionRounded,
+          ),
+        )
+
+        if (!foundTarget.isSelected) {
+          const doubleClick = event.detail > 1 // we interpret a triple click as two double clicks, a quadruple click as three double clicks, etc  // TODO TEST ME
```

Comment:
TODO for later work?

Reply:
yes, all TODOs are meant for follow-up tasks